File: kalitech-engine/src/main/resources/kalitech/engine/modules/Physics/helpers/PhysicsIds.js
"use strict";

function bodyIdOf(handleOrId) {
    if (typeof handleOrId === "number") return handleOrId | 0;
    if (!handleOrId) return 0;
    if (typeof handleOrId.id === "function") return handleOrId.id() | 0;
    if (typeof handleOrId.id === "number") return handleOrId.id | 0;
    if (typeof handleOrId.bodyId === "number") return handleOrId.bodyId | 0;
    return 0;
}

function surfaceIdOf(handleOrId) {
    if (typeof handleOrId === "number") return handleOrId | 0;
    if (!handleOrId) return 0;
    if (typeof handleOrId.id === "function") return handleOrId.id() | 0;
    if (typeof handleOrId.id === "number") return handleOrId.id | 0;
    if (typeof handleOrId.surfaceId === "number") return handleOrId.surfaceId | 0;
    return 0;
}

module.exports = Object.freeze({
    bodyIdOf,
    surfaceIdOf,
});

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Physics/helpers/PhysicsIds.lua
local M = {}
function bodyIdOf(self, handleOrId)
    if KTypeOf(handleOrId) == "number" then
        return bit32.bor(handleOrId, 0)
    end
    if not handleOrId then
        return 0
    end
    if KTypeOf(handleOrId.id) == "function" then
        return bit32.bor(
            handleOrId:id(),
            0
        )
    end
    if KTypeOf(handleOrId.id) == "number" then
        return bit32.bor(handleOrId.id, 0)
    end
    if KTypeOf(handleOrId.bodyId) == "number" then
        return bit32.bor(handleOrId.bodyId, 0)
    end
    return 0
end
function surfaceIdOf(self, handleOrId)
    if KTypeOf(handleOrId) == "number" then
        return bit32.bor(handleOrId, 0)
    end
    if not handleOrId then
        return 0
    end
    if KTypeOf(handleOrId.id) == "function" then
        return bit32.bor(
            handleOrId:id(),
            0
        )
    end
    if KTypeOf(handleOrId.id) == "number" then
        return bit32.bor(handleOrId.id, 0)
    end
    if KTypeOf(handleOrId.surfaceId) == "number" then
        return bit32.bor(handleOrId.surfaceId, 0)
    end
    return 0
end
M = KObject:freeze({bodyIdOf = bodyIdOf, surfaceIdOf = surfaceIdOf})

return M
